File: lib/biometrics/ElementalCoherenceCalculator.ts
/**
 * Elemental Coherence Calculator
 *
 * Maps biometric data to Air/Fire/Water/Earth/Aether qualities
 * Provides unified field coherence scoring for MAIA-PAI system
 *
 * All parameters externalized to /lib/config/BiometricParameters.ts
 * following Spiralogic principle: clarity through explicit constants
 */

import type { ParsedHealthData, HRVReading } from './HealthDataImporter';
import type { CoherenceState } from './CoherenceDetector';
import { BIOMETRIC_CONFIG } from '../config/BiometricParameters';

export interface ElementalCoherence {
  air: number;    // 0-1 (clarity, adaptability, breath)
  fire: number;   // 0-1 (activation, transformation, energy)
  water: number;  // 0-1 (flow, rhythm, emotional regulation)
  earth: number;  // 0-1 (grounding, stability, embodiment)
  aether: number; // 0-1 (unity, peak coherence, integration)
  unified: number; // 0-1 (overall field coherence)
}

export interface BiometricSnapshot {
  hrv: number;              // milliseconds (SDNN)
  heartRate: number;        // BPM
  restingHeartRate: number; // BPM
  respiratoryRate: number;  // breaths/min
  sleepHours: number;
  deepSleepHours: number;
  hrvTrend: 'rising' | 'stable' | 'falling';
  hrvVariance: number;      // statistical variance
  readinessScore: number;   // 0-100
}

export interface KairosWindow {
  isOpen: boolean;
  strength: number; // 0-1
  duration: number; // minutes remaining
  recommendation: string;
}

export class ElementalCoherenceCalculator {

  /**
   * Calculate elemental coherence from biometric snapshot
   */
  calculate(data: BiometricSnapshot): ElementalCoherence {
    const air = this.calculateAir(data);
    const fire = this.calculateFire(data);
    const water = this.calculateWater(data);
    const earth = this.calculateEarth(data);

    // Aether requires all elements to calculate balance
    const aether = this.calculateAether(data, { air, fire, water, earth });

    // Unified field coherence
    const unified = this.calculateUnified({ air, fire, water, earth, aether });

    return { air, fire, water, earth, aether, unified };
  }

  /**
   * Calculate from parsed health data (convenience method)
   */
  calculateFromHealthData(data: ParsedHealthData, coherence: CoherenceState): ElementalCoherence {
    const snapshot = this.healthDataToSnapshot(data, coherence);
    return this.calculate(snapshot);
  }

  /**
   * AIR - Clarity, adaptability, nervous system flexibility
   */
  private calculateAir(data: BiometricSnapshot): number {
    const { HRV, HRV_VARIANCE, BREATHING, WEIGHTS } = BIOMETRIC_CONFIG;

    // Higher HRV = more adaptability
    const hrvScore = this.normalize(data.hrv, HRV.MIN, HRV.MAX);

    // Optimal breathing: coherence frequency
    const breathScore = this.bellCurve(
      data.respiratoryRate,
      BREATHING.AIR.OPTIMAL,
      BREATHING.AIR.VARIANCE,
      BREATHING.AIR.TOLERANCE
    );

    // Higher variance = more flexibility
    const varianceScore = this.normalize(data.hrvVariance, HRV_VARIANCE.MIN, HRV_VARIANCE.MAX);

    return this.weightedAverage([
      { value: hrvScore, weight: WEIGHTS.AIR.HRV },
      { value: breathScore, weight: WEIGHTS.AIR.BREATHING },
      { value: varianceScore, weight: WEIGHTS.AIR.VARIANCE }
    ]);
  }

  /**
   * FIRE - Activation, transformation, vitality
   */
  private calculateFire(data: BiometricSnapshot): number {
    const { READINESS, HEART_RATE, TRENDS } = BIOMETRIC_CONFIG;

    // Base: Readiness score
    let fireScore = data.readinessScore / READINESS.SCALE;

    // Activation bonus: Higher heart rate = more fire
    if (data.heartRate > HEART_RATE.ACTIVE.BASELINE) {
      fireScore = Math.min(1, fireScore + HEART_RATE.ACTIVE.ACTIVATION_BONUS);
    }

    // Transformation bonus: Rising HRV = active transformation
    if (data.hrvTrend === 'rising') {
      fireScore = Math.min(1, fireScore + TRENDS.RISING.FIRE_BONUS);
    }

    return this.clamp(fireScore, 0, 1);
  }

  /**
   * WATER - Flow, emotional regulation, rhythmic consistency
   */
  private calculateWater(data: BiometricSnapshot): number {
    const { HRV_VARIANCE, BREATHING, TRENDS, WEIGHTS } = BIOMETRIC_CONFIG;

    // Flow state: Stable or gently falling HRV
    const flowBonus = data.hrvTrend === 'stable' ? TRENDS.STABLE.WATER_BONUS :
                      data.hrvTrend === 'falling' ? TRENDS.FALLING.WATER_BONUS : 0;

    // Rhythm score: Low variance = steady flow (inverted)
    const rhythmScore = 1 - this.normalize(data.hrvVariance, HRV_VARIANCE.MIN, HRV_VARIANCE.MAX);

    // Slow breathing = more flow
    const breathFlow = this.bellCurve(
      data.respiratoryRate,
      BREATHING.WATER.OPTIMAL,
      BREATHING.WATER.VARIANCE,
      BREATHING.WATER.TOLERANCE
    );

    return this.weightedAverage([
      { value: rhythmScore, weight: WEIGHTS.WATER.RHYTHM },
      { value: breathFlow, weight: WEIGHTS.WATER.BREATHING },
      { value: flowBonus, weight: WEIGHTS.WATER.TREND }
    ]);
  }

  /**
   * EARTH - Grounding, stability, embodiment
   */
  private calculateEarth(data: BiometricSnapshot): number {
    const { HEART_RATE, SLEEP, HRV_VARIANCE, WEIGHTS } = BIOMETRIC_CONFIG;

    // Lower resting heart rate = more grounded (inverted)
    const rhrScore = 1 - this.normalize(
      data.restingHeartRate,
      HEART_RATE.RESTING.MIN,
      HEART_RATE.RESTING.MAX
    );

    // Optimal sleep: bell curve around 7.5 hours
    const sleepScore = this.bellCurve(
      data.sleepHours,
      SLEEP.TOTAL.OPTIMAL,
      SLEEP.TOTAL.VARIANCE,
      SLEEP.TOTAL.TOLERANCE
    );

    // Deep sleep quality
    const deepSleepScore = this.normalize(
      data.deepSleepHours,
      SLEEP.DEEP_SLEEP.MIN,
      SLEEP.DEEP_SLEEP.MAX
    );

    // Stability: Low HRV variance = grounded
    const stabilityScore = 1 - this.normalize(data.hrvVariance, HRV_VARIANCE.MIN, HRV_VARIANCE.MAX);

    return this.weightedAverage([
      { value: rhrScore, weight: WEIGHTS.EARTH.RHR },
      { value: sleepScore, weight: WEIGHTS.EARTH.SLEEP },
      { value: deepSleepScore, weight: WEIGHTS.EARTH.DEEP_SLEEP },
      { value: stabilityScore, weight: WEIGHTS.EARTH.STABILITY }
    ]);
  }

  /**
   * AETHER - Unity, transcendence, peak coherence, integration
   */
  private calculateAether(
    data: BiometricSnapshot,
    elements: { air: number; fire: number; water: number; earth: number }
  ): number {
    const { WEIGHTS, READINESS } = BIOMETRIC_CONFIG;

    // Peak HRV bonus
    const peakHrvBonus = data.hrv > WEIGHTS.AETHER.HRV_PEAK_HIGH ? 0.2 :
                        data.hrv > WEIGHTS.AETHER.HRV_PEAK_MEDIUM ? 0.1 : 0;

    // System integration (readiness)
    const integrationScore = data.readinessScore / READINESS.SCALE;

    // Elemental balance: All elements in harmony
    const balanceScore = this.calculateElementalBalance(elements);

    return this.weightedAverage([
      { value: integrationScore, weight: WEIGHTS.AETHER.PEAK_COHERENCE },
      { value: balanceScore, weight: WEIGHTS.AETHER.ELEMENT_BALANCE },
      { value: peakHrvBonus, weight: WEIGHTS.AETHER.PEAK_BONUS }
    ]);
  }

  /**
   * UNIFIED FIELD - Overall coherence
   */
  private calculateUnified(elements: ElementalCoherence): number {
    const { WEIGHTS } = BIOMETRIC_CONFIG;

    // Weight Aether highest (integration), then balance of all four elements
    return this.weightedAverage([
      { value: elements.aether, weight: WEIGHTS.UNIFIED.AETHER },
      { value: elements.air, weight: WEIGHTS.UNIFIED.AIR },
      { value: elements.fire, weight: WEIGHTS.UNIFIED.FIRE },
      { value: elements.water, weight: WEIGHTS.UNIFIED.WATER },
      { value: elements.earth, weight: WEIGHTS.UNIFIED.EARTH }
    ]);
  }

  /**
   * Calculate elemental balance (1.0 = perfect balance, 0 = extreme imbalance)
   */
  private calculateElementalBalance(elements: { air: number; fire: number; water: number; earth: number }): number {
    const { WEIGHTS } = BIOMETRIC_CONFIG;

    const values = [elements.air, elements.fire, elements.water, elements.earth];
    const mean = values.reduce((sum, v) => sum + v, 0) / values.length;

    // Calculate standard deviation
    const variance = values.reduce((sum, v) => sum + Math.pow(v - mean, 2), 0) / values.length;
    const stdDev = Math.sqrt(variance);

    // Lower standard deviation = better balance
    // Convert to 0-1 scale using configured max stddev
    return Math.max(0, 1 - (stdDev / WEIGHTS.BALANCE.MAX_STDDEV));
  }

  /**
   * Detect Kairos windows (optimal moments for transformation)
   */
  detectKairosWindow(
    elemental: ElementalCoherence,
    coherence: CoherenceState
  ): KairosWindow {
    const { KAIROS } = BIOMETRIC_CONFIG;

    // Kairos requires:
    // 1. High unified coherence
    // 2. Aether above threshold
    // 3. All elements balanced (no single element below minimum)
    // 4. Rising or stable trend

    const isCoherenceHigh = elemental.unified > KAIROS.COHERENCE_MIN;
    const isAetherHigh = elemental.aether > KAIROS.AETHER_MIN;
    const areElementsBalanced =
      elemental.air > KAIROS.ELEMENT_MIN &&
      elemental.fire > KAIROS.ELEMENT_MIN &&
      elemental.water > KAIROS.ELEMENT_MIN &&
      elemental.earth > KAIROS.ELEMENT_MIN;
    const isTrendFavorable = coherence.trend !== 'falling';

    const isOpen = isCoherenceHigh && isAetherHigh && areElementsBalanced && isTrendFavorable;

    // Calculate strength (0-1)
    const strength = isOpen ?
      Math.min(1, elemental.unified * elemental.aether) : 0;

    // Estimate duration based on trend and coherence
    // Stable/rising = longer windows
    let duration = 0;
    if (isOpen) {
      duration = coherence.trend === 'stable' ? KAIROS.DURATION_STABLE :
                 coherence.trend === 'rising' ? KAIROS.DURATION_RISING :
                 KAIROS.DURATION_DEFAULT;

      // Adjust by strength
      duration = Math.round(duration * strength);
    }

    // Generate recommendation
    let recommendation = '';
    if (isOpen) {
      if (strength > KAIROS.STRENGTH_EXCELLENT) {
        recommendation = 'Peak coherence for deep transformational work. Enter Scribe mode.';
      } else if (strength > KAIROS.STRENGTH_GOOD) {
        recommendation = 'Strong coherence for focused creative work or shadow integration.';
      } else {
        recommendation = 'Good coherence for meditation, journaling, or contemplative practice.';
      }
    } else if (elemental.unified > 0.6) {
      recommendation = 'Building coherence. Continue with Patient mode work.';
    } else {
      recommendation = 'Rest and restoration recommended. Dialogue mode for gentle integration.';
    }

    return {
      isOpen,
      strength,
      duration,
      recommendation
    };
  }

  /**
   * Get human-readable elemental description
   */
  getElementalDescription(elemental: ElementalCoherence): {
    primary: string;
    secondary: string;
    state: string;
  } {
    // Find primary and secondary elements
    const elements = [
      { name: 'Air', value: elemental.air, emoji: '💨' },
      { name: 'Fire', value: elemental.fire, emoji: '🔥' },
      { name: 'Water', value: elemental.water, emoji: '🌊' },
      { name: 'Earth', value: elemental.earth, emoji: '🌍' }
    ];

    elements.sort((a, b) => b.value - a.value);

    const primary = `${elements[0].emoji} ${elements[0].name}`;
    const secondary = `${elements[1].emoji} ${elements[1].name}`;

    // Overall state based on unified coherence
    let state = '';
    if (elemental.unified > 0.85) {
      state = 'Peak Coherence - All elements unified';
    } else if (elemental.unified > 0.75) {
      state = 'High Coherence - Strong integration';
    } else if (elemental.unified > 0.6) {
      state = 'Good Coherence - Building';
    } else if (elemental.unified > 0.4) {
      state = 'Medium Coherence - Balancing';
    } else {
      state = 'Low Coherence - Rest needed';
    }

    return { primary, secondary, state };
  }

  // ========== UTILITY METHODS ==========

  /**
   * Normalize value to 0-1 range
   */
  private normalize(value: number, min: number, max: number): number {
    return this.clamp((value - min) / (max - min), 0, 1);
  }

  /**
   * Bell curve centered on target
   */
  private bellCurve(value: number, target: number, spread: number, peak: number = 1): number {
    const distance = Math.abs(value - target);
    const normalized = Math.exp(-Math.pow(distance / spread, 2));
    return normalized * peak;
  }

  /**
   * Clamp value between min and max
   */
  private clamp(value: number, min: number, max: number): number {
    return Math.min(max, Math.max(min, value));
  }

  /**
   * Weighted average
   */
  private weightedAverage(items: Array<{ value: number; weight: number }>): number {
    const totalWeight = items.reduce((sum, item) => sum + item.weight, 0);
    const weightedSum = items.reduce((sum, item) => sum + item.value * item.weight, 0);
    return weightedSum / totalWeight;
  }

  /**
   * Convert ParsedHealthData to BiometricSnapshot
   */
  private healthDataToSnapshot(data: ParsedHealthData, coherence: CoherenceState): BiometricSnapshot {
    // Get latest values
    const latestHRV = data.hrv[0]?.value || 50;
    const latestHR = data.heartRate[0]?.value || 70;
    const restingHR = data.heartRate.find(r => r.context === 'resting')?.value || 65;
    const latestResp = data.respiratory[0]?.value || 15;

    // Calculate sleep
    const totalSleep = data.sleep.reduce((sum, s) => sum + s.durationHours, 0);
    const deepSleep = data.sleep
      .filter(s => s.stage === 'deep')
      .reduce((sum, s) => sum + s.durationHours, 0);

    // Calculate HRV variance
    const hrvValues = data.hrv.slice(0, 20).map(r => r.value); // Last 20 readings
    const hrvMean = hrvValues.reduce((sum, v) => sum + v, 0) / hrvValues.length;
    const hrvVariance = hrvValues.reduce((sum, v) => sum + Math.pow(v - hrvMean, 2), 0) / hrvValues.length;

    return {
      hrv: latestHRV,
      heartRate: latestHR,
      restingHeartRate: restingHR,
      respiratoryRate: latestResp,
      sleepHours: totalSleep,
      deepSleepHours: deepSleep,
      hrvTrend: coherence.trend,
      hrvVariance,
      readinessScore: data.readinessScore
    };
  }
}

// Export singleton
export const elementalCoherenceCalculator = new ElementalCoherenceCalculator();

File: ios-app/MAIA/Services/ElementalCoherenceCalculator.swift
//
//  ElementalCoherenceCalculator.swift
//  MAIA
//
//  Calculates elemental coherence from biometric data
//  Ported from TypeScript - same algorithms!
//

import Foundation

class ElementalCoherenceCalculator {

    // MARK: - Main Calculation

    func calculate(from snapshot: BiometricSnapshot) -> ElementalCoherence {
        let air = calculateAir(from: snapshot)
        let fire = calculateFire(from: snapshot)
        let water = calculateWater(from: snapshot)
        let earth = calculateEarth(from: snapshot)
        let aether = calculateAether(from: snapshot, elements: (air, fire, water, earth))
        let unified = calculateUnified(air: air, fire: fire, water: water, earth: earth, aether: aether)

        return ElementalCoherence(
            air: air,
            fire: fire,
            water: water,
            earth: earth,
            aether: aether,
            unified: unified,
            timestamp: Date()
        )
    }

    // MARK: - AIR: Clarity, Adaptability, Breath

    private func calculateAir(from snapshot: BiometricSnapshot) -> Double {
        // HRV component (nervous system flexibility) - 60%
        let hrvScore = normalize(snapshot.hrv, min: 20, max: 100)

        // Respiratory rate (optimal around 12-18) - 20%
        let respDeviation = abs(snapshot.respiratoryRate - 15.0)
        let respScore = normalize(15 - respDeviation, min: 0, max: 15)

        // HRV variance (lower = more stable = better) - 20%
        let varianceScore = 1.0 - normalize(snapshot.hrvVariance, min: 50, max: 500)

        return clamp(hrvScore * 0.6 + respScore * 0.2 + varianceScore * 0.2)
    }

    // MARK: - FIRE: Transformation, Activation, Energy

    private func calculateFire(from snapshot: BiometricSnapshot) -> Double {
        // Readiness score (activation potential) - 70%
        let readinessScore = normalize(snapshot.readinessScore, min: 0, max: 100)

        // HRV trend bonus (rising = more fire) - 20%
        let trendBonus: Double = {
            switch snapshot.hrvTrend {
            case .rising: return 0.2
            case .stable: return 0.0
            case .falling: return -0.1
            }
        }()

        // Peak state bonus - 10%
        let peakBonus = snapshot.isPeakState ? 0.1 : 0.0

        return clamp(readinessScore * 0.7 + trendBonus + peakBonus)
    }

    // MARK: - WATER: Flow, Rhythm, Emotional Regulation

    private func calculateWater(from snapshot: BiometricSnapshot) -> Double {
        // HRV stability (lower variance = better flow) - 60%
        let stabilityScore = 1.0 - normalize(snapshot.hrvVariance, min: 50, max: 500)

        // Rhythm consistency (heart rate regularity) - 40%
        // Using resting HR as proxy: closer to optimal (50-65) = better
        let rhythmScore = 1.0 - normalize(abs(snapshot.restingHeartRate - 57.5), min: 0, max: 20)

        return clamp(stabilityScore * 0.6 + rhythmScore * 0.4)
    }

    // MARK: - EARTH: Grounding, Stability, Embodiment

    private func calculateEarth(from snapshot: BiometricSnapshot) -> Double {
        // Sleep quality (total + deep) - 50%
        let sleepScore = min(snapshot.sleepHours / 8.0, 1.0) * 0.7
        let deepSleepBonus = min(snapshot.deepSleepHours / 2.0, 1.0) * 0.3

        // Resting heart rate (lower = more grounded) - 30%
        let rhrScore = 1.0 - normalize(snapshot.restingHeartRate, min: 50, max: 85)

        // Embodiment baseline - 20%
        let embodimentBase = 0.5 // Default grounded state

        return clamp((sleepScore + deepSleepBonus) * 0.5 + rhrScore * 0.3 + embodimentBase * 0.2)
    }

    // MARK: - AETHER: Unity, Integration, Transcendence

    private func calculateAether(from snapshot: BiometricSnapshot, elements: (air: Double, fire: Double, water: Double, earth: Double)) -> Double {
        // Elemental balance (all elements present) - 50%
        let allElements = [elements.air, elements.fire, elements.water, elements.earth]
        let maxElement = allElements.max() ?? 0
        let minElement = allElements.min() ?? 0
        let balance = 1.0 - (maxElement - minElement)

        // Peak state HRV (very high HRV = transcendent moments) - 30%
        let peakScore = normalize(snapshot.hrv, min: 50, max: 100)

        // Integration bonus (readiness + balance) - 20%
        let integrationScore = normalize(snapshot.readinessScore, min: 0, max: 100) * balance

        return clamp(balance * 0.5 + peakScore * 0.3 + integrationScore * 0.2)
    }

    // MARK: - UNIFIED: Overall Field Coherence

    private func calculateUnified(air: Double, fire: Double, water: Double, earth: Double, aether: Double) -> Double {
        return (air + fire + water + earth + aether) / 5.0
    }

    // MARK: - Utility Functions

    /// Normalize value between 0 and 1
    private func normalize(_ value: Double, min minValue: Double, max maxValue: Double) -> Double {
        guard maxValue > minValue else { return 0.5 }
        return clamp((value - minValue) / (maxValue - minValue))
    }

    /// Clamp value between 0 and 1
    private func clamp(_ value: Double) -> Double {
        return max(0.0, min(1.0, value))
    }
}

// MARK: - Singleton

extension ElementalCoherenceCalculator {
    static let shared = ElementalCoherenceCalculator()
}
